ocaml/xapi/test_vm.ml

```diff
@@ -0,0 +1,80 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Stdext
+open OUnit
+open Test_highlevel
+
+module VMSetBiosStrings = Generic.Make (Generic.EncapsulateState (struct
+  module Io = struct
+    type input_t = string * (API.vm_bios_string_keys * string) list
+    type output_t = (exn, unit) Either.t
+    let string_of_input_t = fun (a, b) -> Printf.sprintf "%s, %s" a
+      (String.concat "; " (List.map (fun (k,v) -> (Record_util.vm_bios_key_to_string k) ^ "=" ^ v) b))
+    let string_of_output_t = Test_printers.(either exn unit)
+  end
+  module State = Test_state.XapiDb
+
+  let load_input __context (name_label, _) =
+    ignore (Test_common.make_vm ~__context ~name_label ())
+
+  let extract_output __context (label, value) =
+    let self = List.hd (Db.VM.get_by_name_label ~__context ~label) in
+    try Either.Right (Xapi_vm.set_bios_strings ~__context ~self ~value) with e -> Either.Left e
+
+  let big_str = String.make (Xapi_globs.bios_string_limit_size + 1) 'x'
+  let non_printable_str1 = Printf.sprintf "xyz%c" (Char.chr 31)
+  let non_printable_str2 = Printf.sprintf "xyz%c" (Char.chr 127)
+
+  let tests = [
+    (* Empty value *)
+    ("a", [`enclosureassettag, ""]),
+    Either.Left Api_errors.(Server_error
+      (invalid_value,
+      ["enclosure-asset-tag"; "Value provided is empty"]));
+
+    (* Value having more than 512 charactors *)
+    ("a", [`enclosureassettag, big_str]),
+    Either.Left Api_errors.(Server_error
+      (invalid_value,
+      ["enclosure-asset-tag"; (Printf.sprintf "%s has length more than %d characters" big_str Xapi_globs.bios_string_limit_size)]));
+
+    (* Value having non printable ascii characters *)
+    ("a", [`enclosureassettag, non_printable_str1]),
+    Either.Left Api_errors.(Server_error
+      (invalid_value,
+      ["enclosure-asset-tag"; non_printable_str1 ^ " has non-printable ASCII characters"]));
+
+    ("a", [`enclosureassettag, non_printable_str2]),
+    Either.Left Api_errors.(Server_error
+      (invalid_value,
+      ["enclosure-asset-tag"; non_printable_str2 ^ " has non-printable ASCII characters"]));
+
+    (* Correct value *)
+    ("a", [`biosvendor, "Test";
+           `biosversion, "Test Inc. A08";
+           `systemmanufacturer, "Test Inc.";
+           `systemproductname, "Test bios strings";
+           `systemversion, "8.1.1 SP1 build 8901";
+           `systemserialnumber, "test-test-test-test";
+           `enclosureassettag, "testassettag"]),
+    Either.Right ();
```

Comment:
If you change `extract_output` as I said above, then you can add a couple of further tests to check the output. It's important to verify the behaviour when just one or a few strings a specified (most notably the new asset tag one), and check that they are supplemented with the standard ones in the right way.

Reply:
Update, Thanks @robhoes